### PR TITLE
feat: add refresh token support to nodejs sdk

### PR DESCRIPTION
**Pull request by ben-kugler**
- add refresh token and generate token api's
- made test ignore auth.test.ts as these test will require a session token, no easy way right now to programmatically get a session token atm

Test run fine locally, hitting auth issues here
```
 npm ERR! code E401
npm ERR! Unable to authenticate, your authentication token seems to be invalid.
npm ERR! To correct this please trying logging in again with:
npm ERR!     npm login
```
image

### CODE REVIEW

File: CONTRIBUTING.md
Comment (cprice404) [inline CONTRIBUTING.md:69]: @bruuuuuuuce @rishtigupta next week we should talk about our options for making it possible to run these in CI

File: packages/client-sdk-nodejs/src/auth-client-props.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/src/auth-client-props.ts:12]: I'm 90% sure that we don't want this field in our final public API for this.  That credential provider is for providing API tokens, which AFAIK will not be used for many/most of the APIs that people consume the AuthClient for.

Once this PR is merged we will either need to be happy committing to what is in it as a public API for AuthClient, since it would be included in the next node.js release, or we will be blocked on doing any more node.js releases until we file follow-up PRs to solidify the API.  My pref would be to try to get it all sorted in this PR - apologies that this is landing on you Ben, we were planning to do this stuff but you started working on this before we had time.

Comment (cprice404) [reply packages/client-sdk-nodejs/src/auth-client-props.ts:12]: cc: @rlinehan

File: packages/client-sdk-nodejs/src/auth-client.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/src/auth-client.ts:34]: this is another thing that I'm not sure I want in our public API, see: 

https://github.com/momentohq/client-sdk-nodejs/issues/394

File: packages/client-sdk-nodejs/src/auth-client-props.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/src/auth-client-props.ts:18]: do we need to add this deprecated thing? :)

File: packages/client-sdk-nodejs/src/index.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/src/index.ts:3]: errant comment?

File: packages/client-sdk-nodejs/src/internal/auth-client.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/src/internal/auth-client.ts:39]: This shouldn't be necessary for the Auth APIs, right?

Comment (cprice404) [reply packages/client-sdk-nodejs/src/internal/auth-client.ts:39]: cc: @rlinehan @bruuuuuuuce

File: packages/client-sdk-nodejs/test/integration/auth.test.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/test/integration/auth.test.ts:36]: The structure of this `success` response is another thing we need to think about the API for.  I suspect that we will want to use the same `ExpiresAt` or `ExpiresIn` type of object as I linked to in one of my comments above, as opposed to having `undefined` imply "never expires".  We can discuss this if people have thoughts but we have to get it right before we ship this.

File: packages/client-sdk-nodejs/test/integration/auth.test.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/test/integration/auth.test.ts:8]: i know you are just copying an existing pattern here but it's a weird pattern that I want to clean up.  Can you make these names start with a lower-case letter, and instead of `Momento` can you call this `authClient`?  I would also love to see type annotations on these two variables to make it more obvious what they are.

File: packages/client-sdk-nodejs/test/integration/auth.test.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/test/integration/auth.test.ts:80]: supernit but there should be a `delay` function somewhere in this repo that you can just `await` directly, and it makes it a bit more readable than the setTimeout stuff.

File: packages/client-sdk-nodejs/test/integration/integration-setup.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/test/integration/integration-setup.ts:49]: can we annotate return type here please

File: packages/client-sdk-nodejs/test/integration/integration-setup.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/test/integration/integration-setup.ts:84]: same comment about the names of these vars

File: packages/core/src/messages/responses/refresh-api-token.ts
Comment (cprice404) [inline packages/core/src/messages/responses/refresh-api-token.ts:38]: explicity return type annotations on all of these please

**Review by cprice404 — CHANGES_REQUESTED** [review]

File: packages/client-sdk-nodejs/src/internal/auth-client.ts
Comment (rlinehan) [reply packages/client-sdk-nodejs/src/internal/auth-client.ts:39]: yep, shouldn't be necessary

File: packages/client-sdk-nodejs/test/integration/auth.test.ts
Comment (ben-kugler) [reply packages/client-sdk-nodejs/test/integration/auth.test.ts:8]: For sure!

File: packages/client-sdk-nodejs/src/internal/auth-client.ts
Comment (ben-kugler) [reply packages/client-sdk-nodejs/src/internal/auth-client.ts:39]: No it shouldn't, just a copy-paste from another client - removing!

File: packages/client-sdk-nodejs/src/auth-client-props.ts
Comment (cprice404) [reply packages/client-sdk-nodejs/src/auth-client-props.ts:12]: this is probably the biggest UX/API issue we need to sort out.  We need the control plane endpoint for this client, but not all of the APIs in this client require an API token.  We don't currently have a mechanism for a user to get access to the control plane endpoint without parsing it from an API token.

From a practical perspective this client should only require the endpoint uri to its constructor, but from a UX perspective I am on the fence about how they would discover that and get it into their code.

Comment (ben-kugler) [reply packages/client-sdk-nodejs/src/auth-client-props.ts:12]: We have the same issue for momento cli atm, we plan to just hardcode endpoints for now w/ custom option. Seems like we don't have a real solution without a endpoint discovery system, which it seems we are going to need really soon here. 

For now, seems best that we supply a helper `(aws/gcp, region): endpoint` in node sdk and have the `AuthClient` only accept a url.

File: packages/client-sdk-nodejs/package-lock.json
Comment (bruuuuuuuce) [inline packages/client-sdk-nodejs/package-lock.json:12538]: this is why you are getting an npm error. These are dependencies installed via our internal/private npm registries. We should point all deps to the public npm registry. To fix this, you can remove this file, remove your `~/.npmrc` file, and `npm i` to reinstall all deps using the public registry

**Comment by ben-kugler** [comment]
Newest build fails, after fighting the linter this error I'm not entirely sure how to correct as they appear to be coming from the generated node modules.

**Review by cprice404 — CHANGES_REQUESTED** [review]

File: packages/core/src/utils/expiration.ts
Comment (ben-kugler) [reply packages/core/src/utils/expiration.ts]: Yeah, just forgot to remove that, updated!

**Comment by cprice404** [comment]
@ben-kugler if you update the generated-types dep to the one that just got released from Matt's client-protos PR, I think it will fix the build issue.

File: packages/core/test/unit/utils/expiration.test.ts
Comment (cprice404) [inline packages/core/test/unit/utils/expiration.test.ts]: this shouldn't compile with your latest changes should it?

File: packages/client-sdk-nodejs/src/internal/auth-client.ts
Comment (ben-kugler) [reply packages/client-sdk-nodejs/src/internal/auth-client.ts]: It is being used in the construction of `IdleGrpcClientWrapper` in `generateApiToken`/`refreshApiToken` calls and then we also log a little but here. 

I'm also now seeing errors from the `client-sdk-web` due to changes in the shared code, so guess I'm making some updates there to match the `client-node-sdk` changes.

Comment (cprice404) [reply packages/client-sdk-nodejs/src/internal/auth-client.ts]: ah, I see.  You don't need that IdleGrpcClientWrapper for now, you can just use the grpc.AuthClient object directly.  So let's get rid of that and then I think my recommendation for now would be to remove the log statements and remove this props object.  That gives us the most flexibility in the future without committing to a specific API for config/logging settings for now.  lmk if that causes any trouble

File: packages/client-sdk-nodejs/src/auth-client-props.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/src/auth-client-props.ts]: is this still used?

File: packages/client-sdk-nodejs/src/auth-client.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/src/auth-client.ts]: I was thinking we could ditch this here in the public API for now, not just the internal one, sorry if that wasn't clear.

**Review by cprice404 — COMMENTED** [review]
i don't see anything else left other than the AuthClientProps thing, I think I will be +1 as soon as we take that out

**Review by cprice404 — APPROVED** [review]

**Pull request merged**
